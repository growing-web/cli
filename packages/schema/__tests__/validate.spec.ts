import type { WebBuilderManifest } from '@growing-web/web-builder-types'

import { test, describe, expect } from 'vitest'
import { validate } from '@growing-web/web-builder-schema'

describe('schema validate test.', () => {
  /**
   * 捕获校验错误进行测试
   * @param type
   * @param config
   */
  const validateSchema = (type: string, config: WebBuilderManifest) => {
    let error: any
    try {
      validate(config)
    } catch (errorFromPlugin: any) {
      if (errorFromPlugin.name !== 'ValidationError') {
        throw error
      }
      error = errorFromPlugin
    } finally {
      if (type === 'success') {
        expect(error).toBeUndefined()
      } else if (type === 'failure') {
        expect(() => {
          throw error
        }).toThrowErrorMatchingSnapshot()
      }
    }
  }

  const tests: Record<'success' | 'failure', WebBuilderManifest[]> = {
    success: [
      {
        schemaVersion: '0.0.0',
        entry: 'app.js',
      },
      {
        schemaVersion: '0.0.0',
        entry: 'src/main.ts',
      },
      {
        schemaVersion: '0.0.0',
        entry: 'src/main.ts',
        externals: {
          Vue: 'Vue',
        },
      },
      {
        schemaVersion: '0.0.0',
        entry: 'src/main.ts',
        publicPath: '/',
        externals: {},
        server: { port: 3000, proxy: [], https: false, host: '' },
      },
      {
        schemaVersion: '0.0.0',
        entry: 'src/main.ts',
        publicPath: '/',
        externals: { lazy: 'React' },
        server: { port: 3000, proxy: [], https: false, host: '' },
      },
      {
        schemaVersion: '0.0.0',
        entry: 'app.js',
        server: { proxy: [] },
      },
      {
        schemaVersion: '0.0.0',
        entry: 'app.js',
        server: { proxy: [{ url: '/api', target: 'https://xxx' }] },
      },
      {
        schemaVersion: '0.0.0',
        entry: 'app.js',
        server: {
          proxy: [
            {
              url: '/api',
              target: 'https://xxx',
              pathRewrite: [{ find: /^\/api/, replacement: '' }],
            },
            {
              url: '/api',
              target: 'https://xxx',
              pathRewrite: [{ find: '/api', replacement: '' }],
            },
          ],
        },
      },
      {
        schemaVersion: '0.0.0',
        entry: 'app.js',
        externals: { vue: 'Vue' },
      },
      {
        schemaVersion: '0.0.0',
        entry: 'app.js',
        env: {
          development: {
            publicPath: '/',
          },
        },
      },
      {
        schemaVersion: '0.0.0',
        entry: 'app.js',
        env: {
          development: {
            server: { port: 4000 },
          },
        },
      },
    ],
    failure: [
      {
        schemaVersion: 1,
        entry: 'app.js',
      },
      {
        schemaVersion: '0.0.0',
        entry: 'src/main.ts',
        externals: {
          123: 'Vue',
        },
      },
      {
        entry: 'src/main.ts',
      },
      {
        schemaVersion: '0.0.0',
        externals: '',
      },
      {
        schemaVersion: '0.0.0',
        entry: 'app.js',
        server: { proxy: [''] },
      },
      {
        schemaVersion: '0.0.0',
        entry: 'app.js',
        server: { https: 'x' },
      },
      {
        schemaVersion: '0.0.0',
        entry: 'app.js',
        server: { host: 100 },
      },
      {
        schemaVersion: '0.0.0',
        entry: 'app.js',
        server: { port: '3000' },
      },
      {
        schemaVersion: '0.0.0',
        entry: 'app.js',
        server: { proxy: [{}] },
      },
      {
        schemaVersion: '0.0.0',
        entry: 'app.js',
        server: {
          proxy: [
            {
              url: '/api',
              target: 'https://xxx',
              pathRewrite: [{ find: 123, replacement: '' }],
            },
          ],
        },
      },
      {
        schemaVersion: '0.0.0',
        entry: 'app.js',
        env: {
          123: {
            server: { port: 4000 },
          },
        },
      },
      {
        schemaVersion: '0.0.0',
        entry: 'app.js',
        env: {
          development: {
            server: { port: 99999 },
          },
        },
      },
    ] as any,
  }
  for (const type of Object.keys(tests) as ['success', 'failure']) {
    for (const value of tests[type]) {
      test(`should ${
        type === 'success' ? 'successfully validate' : 'throw an error on'
      } the patterns option`, () => {
        validateSchema(type, value)
      })
    }
  }
})
