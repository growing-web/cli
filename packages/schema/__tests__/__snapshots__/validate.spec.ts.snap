// Vitest Snapshot v1

exports[`schema validate test. > should throw an error on the patterns option 1`] = `
"Invalid Manifest object. Project has been initialized using a Manifest object that does not match the API schema.
 - Manifest.schemaVersion should be a string.
   -> The version number of the manifest."
`;

exports[`schema validate test. > should throw an error on the patterns option 2`] = `
"Invalid Manifest object. Project has been initialized using a Manifest object that does not match the API schema.
 - Manifest.externals has an unknown property '123'. These properties are valid:
   object {} (additional property names should match pattern \\"[a-zA-Z]+\\")"
`;

exports[`schema validate test. > should throw an error on the patterns option 3`] = `
"Invalid Manifest object. Project has been initialized using a Manifest object that does not match the API schema.
 - Manifest misses the property 'schemaVersion'. Should be:
   string
   -> The version number of the manifest."
`;

exports[`schema validate test. > should throw an error on the patterns option 4`] = `
"Invalid Manifest object. Project has been initialized using a Manifest object that does not match the API schema.
 - Manifest misses the property 'entry'. Should be:
   string
 - Manifest.externals should be an object:
   object {} (additional property names should match pattern \\"[a-zA-Z]+\\")"
`;

exports[`schema validate test. > should throw an error on the patterns option 5`] = `
"Invalid Manifest object. Project has been initialized using a Manifest object that does not match the API schema.
 - Manifest.server.proxy[0] should be an object:
   object { url, target, secure?, changeOrigin?, pathRewrite?, â€¦ }"
`;

exports[`schema validate test. > should throw an error on the patterns option 6`] = `
"Invalid Manifest object. Project has been initialized using a Manifest object that does not match the API schema.
 - Manifest.server.https should be a boolean."
`;

exports[`schema validate test. > should throw an error on the patterns option 7`] = `
"Invalid Manifest object. Project has been initialized using a Manifest object that does not match the API schema.
 - Manifest.server.host should be a string."
`;

exports[`schema validate test. > should throw an error on the patterns option 8`] = `
"Invalid Manifest object. Project has been initialized using a Manifest object that does not match the API schema.
 - Manifest.server.port should be a number (should be >= 0 and <= 65535)."
`;

exports[`schema validate test. > should throw an error on the patterns option 9`] = `
"Invalid Manifest object. Project has been initialized using a Manifest object that does not match the API schema.
 - Manifest.server.proxy[0] misses the property 'target'. Should be:
   string"
`;

exports[`schema validate test. > should throw an error on the patterns option 10`] = `
"Invalid Manifest object. Project has been initialized using a Manifest object that does not match the API schema.
 - Manifest.server.proxy[0].pathRewrite[0].find should be one of these:
   RegExp | string
   Details:
    * Manifest.server.proxy[0].pathRewrite[0].find should be an instance of RegExp.
    * Manifest.server.proxy[0].pathRewrite[0].find should be a string."
`;

exports[`schema validate test. > should throw an error on the patterns option 11`] = `
"Invalid Manifest object. Project has been initialized using a Manifest object that does not match the API schema.
 - Manifest.env has an unknown property '123'. These properties are valid:
   object {} (additional property names should match pattern \\"[a-z]+\\")
   -> env configuration for different environments."
`;

exports[`schema validate test. > should throw an error on the patterns option 12`] = `
"Invalid Manifest object. Project has been initialized using a Manifest object that does not match the API schema.
 - Manifest.env.development.server.port should be >= 0 and <= 65535."
`;
